### PR TITLE
Add to/from_data to RobotModel

### PR DESCRIPTION
**Pull request by beverlylytle**
Address #564 

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by tomvanmele** [comment]
i would prefer to remove the slots from the ABC rather than making a separate base class. otherwise it kind of defeats the purpose. the slots definition is actually not needed...

**Comment by beverlylytle** [comment]
> i would prefer to remove the slots from the ABC rather than making a separate base class. otherwise it kind of defeats the purpose. the slots definition is actually not needed...

Ok, so then implementing `to_json` and `from_json` is also desired?

**Comment by tomvanmele** [comment]
yes, but perhaps those don't have to be declared as abstract and could in fact be implemented directly in the base class. if this complicates things for you, perhaps remove the abstract decorators from the json methods for now...

File: src/compas/geometry/_primitives/_primitive.py
Comment (brgcode) [inline src/compas/geometry/_primitives/_primitive.py]: why not derive from `Base` here?

**Review by tomvanmele — COMMENTED** [review]
LGTM in general. haven't looked in detail at the robot stuff though...
could you perhaps add `Base` to the geometry primitive as well, instead of `object`?

**Comment by beverlylytle** [comment]
> could you perhaps add `Base` to the geometry primitive as well, instead of `object`?

That is a possibility, but I was returning it to the way it is on the master branch currently.  If that were to happen, that would mean implementing `to/from_json` for all things that inherit from `Primitive` which would be beyond the scope of this PR.

**Comment by tomvanmele** [comment]
fair enough

**Comment by tomvanmele** [comment]
will leave to @gonzalocasas to review the robot stuff and merge

**Comment by beverlylytle** [comment]
> could you perhaps add `Base` to the geometry primitive as well, instead of `object`?

Wasn't so hard as I thought...

**Comment by tomvanmele** [comment]
cool!
@gonzalocasas ...

File: src/compas/robots/model/link.py
Comment (gonzalocasas) [inline src/compas/robots/model/link.py]: what's the `# !!!` for? because type is a reserved keyword and we're using it anyway? ;)

**Review by gonzalocasas — APPROVED** [review]
LGTM! If possible, I would like to add a few simple unit tests. They don't need to be fancy at all, you can copy&paste some of the ones already existing (ones that have a full urdf as input), do a back and forth conversion `to_data` and then `from_data` and then keep the existing checks, so that you don't need to spend time on the details of what to check inside the URDF.

File: src/compas/robots/model/link.py
Comment (beverlylytle) [reply src/compas/robots/model/link.py]: oh, no, i just use '!!!' as a flag for things i want to double check (easy to search for this string in the code).  originally i wasn't sure of the type of `type`, but now that i know it's a string i can just delete my flag.  not fussed about it being a reserved word.

Comment (gonzalocasas) [reply src/compas/robots/model/link.py]: 😉

**Comment by gonzalocasas** [comment]
image

Great!!!

**Comment by beverlylytle** [comment]
I'm waiting for @yck011522's  actual use case testing to be successful before I merge.

**Pull request merged**
